File: lib/src/main/java/it/scoppelletti/spaceship/http/HttpApplicationException.java
/*
 * Copyright (C) 2017 Dario Scoppelletti, <http://www.scoppelletti.it/>.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package it.scoppelletti.spaceship.http;

import java.io.IOException;
import java.io.Reader;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import com.google.gson.Gson;
import lombok.extern.slf4j.Slf4j;
import okhttp3.ResponseBody;
import retrofit2.HttpException;
import retrofit2.Response;
import it.scoppelletti.spaceship.io.IOExt;

/**
 * HTTP application exception.
 *
 * @since 1.0.0
 */
@Slf4j
public final class HttpApplicationException extends RuntimeException {
    private static final long serialVersionUID = 1;

    /**
     * @serial The localized message.
     */
    private String myMessage;

    /**
     * @serial The HTTP status code. May be {@code 0}.
     */
    private int myStatusCode;

    /**
     * @serial Description of the HTTP status code. May be {@code null}.
     */
    private String myError;

    /**
     * @serial Class of the original exception.
     */
    private String myException;

    /**
     * @serial Path of the resource. May be {@code null}.
     */
    private String myPath;

    /**
     * @serial Timestamp of the exception. Represented by a millisecond value
     *         that is an offset from the <i>Epoch</i>.
     */
    private long myTimestamp;

    /**
     * Constructor.
     *
     * @param builder The instance builder.
     */
    private HttpApplicationException(Builder builder) {
        myMessage = builder.message;
        myStatusCode = builder.statusCode;
        myError = builder.error;
        myException = builder.exception;
        myPath = builder.path;
        myTimestamp = builder.timestamp;
    }

    /**
     * Constructor.
     *
     * @param ex The original exception.
     */
    private HttpApplicationException(HttpException ex) {
        myMessage = ex.getMessage();
        myStatusCode = ex.code();
        myError = ex.message();
        myException = HttpException.class.getName();
        myTimestamp = System.currentTimeMillis();
        initCause(ex);
    }

    /**
     * Converts a Retrofit HTTP exception in a Spaceship exception.
     *
     * <p>The response body describing the HTTP error must mimic the JSON object
     * returned by the default {@code ErrorController} implementation in Spring
     * Boot 1.3.4:</p>
     *
     * <p><table width="100%" border="1" cellpadding="5">
     * <thead>
     * <tr>
     *     <th>Key</th>
     *     <th>Value</th>
     * </tr>
     * </thead>
     * <tbody>
     * <tr>
     *     <td>{@code message}</td>
     *     <td>The localized message.</td>
     * </tr>
     * <tr>
     *     <td>{@code statusCode}</td>
     *     <td>The HTTP status code (optional).</td>
     * </tr>
     * <tr>
     *     <td>{@code error}</td>
     *     <td>Description of the HTTP status code (optional).</td>
     * </tr>
     * <tr>
     *     <td>{@code exception}</td>
     *     <td>Class og the original exception.</td>
     * </tr>
     * <tr>
     *     <td>{@code path}</td>
     *     <td>Path of the resource (optional).</td>
     * </tr>
     * <tr>
     *     <td>{@code timestamp}</td>
     *     <td>The timestamp of the exception. Represented by a millisecond value
     *     that is an offset from the <i>Epoch</i>.</td>
     * </tr>
     * </tbody>
     * </table></p>
     *
     * @param  ex The original exception.
     * @return    The converted exception.
     * @see       <a href="http://projects.spring.io/spring-boot"
     *            target="_blank">Spring Boot</a>
     */
    @NonNull
    public static HttpApplicationException create(@NonNull HttpException ex) {
        Gson converter;
        Response<?> resp;
        ResponseBody body;
        Reader reader = null;
        HttpApplicationException applEx;
        HttpApplicationException.Builder builder;

        if (ex == null) {
            throw new NullPointerException("Argument ex is null.");
        }

        resp = ex.response();
        if (resp == null) {
            // The original exception comes from a deserialization
            return new HttpApplicationException(ex);
        }

        converter = new Gson();
        body = resp.errorBody();

        try {
            reader = body.charStream();
            builder = converter.fromJson(reader,
                    HttpApplicationException.Builder.class);
        } catch (RuntimeException err) {
            myLogger.error("Failed to convert response body.", err);
            builder = new HttpApplicationException.Builder(resp, body);
        } finally {
            reader = IOExt.close(reader);
            body = IOExt.close(body);
        }

        applEx = builder.build();
        applEx.copy(ex);
        applEx.initCause(ex);
        return applEx;
    }

    @Override
    public String getMessage() {
        return toString();
    }

    @Override
    public String getLocalizedMessage() {
        return myMessage;
    }

    /**
     * Applies default values from the properties of a Retrofit HTTP exception.
     *
     * @param ex The original exception.
     */
    private void copy(HttpException ex) {
        if (TextUtils.isEmpty(myMessage)) {
            myMessage = ex.getMessage();
        }

        if (myStatusCode == 0) {
            myStatusCode = ex.code();
        }

        if (TextUtils.isEmpty(myError)) {
            myError = ex.message();
        }

        if (TextUtils.isEmpty(myException)) {
            myException = HttpException.class.getName();
        }

        if (myTimestamp == 0L) {
            myTimestamp = System.currentTimeMillis();
        }
    }

    @Override
    public String toString() {
        return new StringBuilder(getClass().getName())
                .append("(message=")
                .append(myMessage)
                .append(";statusCode=")
                .append(myStatusCode)
                .append(";error=")
                .append(myError)
                .append(";exception=")
                .append(myException)
                .append(";path=")
                .append(myPath)
                .append(";timestamp=")
                .append(myTimestamp)
                .append(")").toString();
    }

    /**
     * Gets the HTTP status code.
     *
     * @return The value. May be {@code 0}.
     * @see    #getError()
     */
    public int getStatusCode() {
        return myStatusCode;
    }

    /**
     * Gets the description the HTTP status code.
     *
     * @return The value. May be {@code null}.
     * @see    #getStatusCode()
     */
    @Nullable
    public String getError() {
        return myError;
    }

    /**
     * Gets the class of the original exception.
     *
     * @return The value.
     */
    public String getException() {
       return myException;
    }

    /**
     * Gets the path of the resource.
     *
     * @return The value. May be {@code null}.
     */
    @Nullable
    public String getPath() {
        return myPath;
    }

    /**
     * Gets the timestamp of the exception.
     *
     * @return The value (milliseconds since the <i>Epoch</i>).
     */
    public long getTimestamp() {
        return myTimestamp;
    }

    /**
     * Builds an {@code HttpApplicationException} instance.
     */
    private static final class Builder {

        /**
         * The localized message.
         */
        String message;

        /**
         * The HTTP status code. May be {@code 0}.
         */
        int statusCode;

        /**
         * Description of the HTTP status code. May be {@code null}.
         */
        String error;

        /**
         * Class of the original exception.
         */
        String exception;

        /**
         * Path of the resource. May be {@code null}.
         */
        String path;

        /**
         * Timestamp of the exception. Represented by a millisecond value that
         * is an offset from the <i>Epoch</i>.
         */
        long timestamp;

        /**
         * Sole constructor.
         */
        Builder() {
        }

        /**
         * Constructor.
         *
         * @param  resp The response.
         * @param  body The response body.
         */
        Builder(Response<?> resp, ResponseBody body) {
            try {
                message = body.string();
            } catch (IOException ex) {
                myLogger.error("Failed to read response body.", ex);
                message = resp.toString(); // raw response
            }

            statusCode = resp.code();
            error = resp.message();
            exception = HttpException.class.getName();
            timestamp = System.currentTimeMillis();
        }

        /**
         * Builds a new {@code HttpApplicationException} instance.
         *
         * @return The new object.
         */
        @NonNull
        HttpApplicationException build() {
            return new HttpApplicationException(this);
        }
    }
}

File: spaceship-http/lib/src/main/kotlin/it/scoppelletti/spaceship/http/HttpApplicationException.kt
/*
 * Copyright (C) 2017-2018 Dario Scoppelletti, <http://www.scoppelletti.it/>.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package it.scoppelletti.spaceship.http

import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.JsonClass
import com.squareup.moshi.Moshi
import it.scoppelletti.spaceship.io.closeQuietly
import it.scoppelletti.spaceship.types.StringExt
import it.scoppelletti.spaceship.types.trimRaw
import mu.KLogger
import mu.KotlinLogging
import okhttp3.ResponseBody
import okio.BufferedSource
import retrofit2.HttpException
import retrofit2.Response
import java.io.IOException

private val logger: KLogger = KotlinLogging.logger {}

/**
 * HTTP application exception.
 *
 * @since 1.0.0
 *
 * @property statusCode HTTP status code.
 * @property error      Description of the HTTP status code.
 * @property exception  Class of the original exception from the server.
 * @property path       Path of the resource.
 * @property timestamp  Timestamp of the original exception from the server
 *                      (milliseconds from the *Epoch*).
 */
public class HttpApplicationException private constructor(
        builder: HttpApplicationException.Builder
): RuntimeException() {

    public val statusCode: Int
    public val error: String?
    public val exception: String
    public val path: String?
    public val timestamp: Long
    private val _message: String

    init {
        _message = builder.message
        statusCode = builder.statusCode
        error = builder.error
        exception = builder.exception
        path = builder.path
        timestamp = builder.timestamp
    }

    override val message: String?
        get() = toString()

    override fun getLocalizedMessage(): String = _message

    override fun toString(): String = """
        |HttpApplicationException(message=$_message,statusCode=$statusCode,
        |error=$error,exception=$exception,path=$path,
        |timestamp=$timestamp)""".trimRaw()

    /**
     * Builds an `HttpApplicationException` instance.
     *
     * @property message    Localized message.
     * @property statusCode HTTP status code.
     * @property error      Description of the HTTP status code.
     * @property exception  Class of the original exception from the server.
     * @property path       Path of the resource.
     * @property timestamp  Timestamp of the original exception from the server
     *                      (milliseconds from the *Epoch*).
     */
    @JsonClass(generateAdapter = true)
    internal class Builder {

        var message: String = StringExt.EMPTY
        var statusCode: Int = 0
        var error: String? = null
        var exception: String = StringExt.EMPTY
        var path: String? = null
        var timestamp: Long = 0L

        /**
         * Applies default values from the properties of a Retrofit HTTP
         * exception.
         *
         * @param source The source exception.
         */
        fun applyDefaults(source: HttpException) {
            if (message.isBlank()) {
                message = source.message.orEmpty()
            }

            if (statusCode == 0) {
                statusCode = source.code()
            }

            if (error.isNullOrBlank()) {
                error = source.message()
            }

            if (exception.isBlank()) {
                exception = source.javaClass.name
            }

            if (timestamp == 0L) {
                timestamp = System.currentTimeMillis()
            }
        }

        /**
         * Builds a new `HttpApplicationException` instance.
         *
         * @return The new object.
         */
        fun build(): HttpApplicationException {
            return HttpApplicationException(this)
        }
    }
}

/**
 * Converts a Retrofit HTTP exception to an `HttpApplicationException`
 * exception.
 *
 * * [Server error response](http://github.com/dscoppelletti/spaceship/wiki/Server-error-response)
 *
 * @receiver Source exception.
 * @return   Converted exception.
 * @since    1.0.0
 */
public fun HttpException.toHttpApplicationException(
): HttpApplicationException =
        makeBuilder(this)
                .apply {
                    applyDefaults(this@toHttpApplicationException)
                }
                .build()
                .apply {
                    initCause(this@toHttpApplicationException)
                }

/**
 * Creates a new `HttpApplicationException.Builder` instance.
 *
 * @param  source Original exception.
 * @return        The new object.
 */
private fun makeBuilder(
        source: HttpException
) : HttpApplicationException.Builder {
    val message: String
    val resp: Response<*>?
    val body: ResponseBody?
    val moshi: Moshi
    val stream: BufferedSource
    val adapter: JsonAdapter<HttpApplicationException.Builder>
    val builder: HttpApplicationException.Builder?

    resp = source.response()
    if (resp == null) {
        // The original exception comes from a deserialization
        return HttpApplicationException.Builder()
    }

    body = resp.errorBody()
    if (body == null) {
        return makeBuilder(source, resp, null)
    }

    // I don't want to consume the original Okio source of the body
    stream = body.source().peek()
    try {
        message = stream.readUtf8()
    } catch (ex: IOException) {
        logger.error("Failed to read response body.", ex)
        // Use raw response
        return makeBuilder(source, resp, null)
    } finally {
        stream.closeQuietly()
    }

    moshi = Moshi.Builder().build()
    adapter = moshi.adapter(HttpApplicationException.Builder::class.java)
            .failOnUnknown().nullSafe()

    builder = try {
        adapter.fromJson(message)
    } catch (ex: Exception) { // IOException|JsonDataException
        logger.error("Failed to convert response body.", ex)
        null
    }

    return builder ?: makeBuilder(source, resp, message)
}

/**
 * Creates a new `HttpApplicationException.Builder` instance.
 *
 * @param  source Original exception.
 * @param  resp   Response.
 * @param  body   Response body.
 * @return        The new object.
 */
private fun makeBuilder(
        source: HttpException,
        resp: Response<*>,
        body: String?
) : HttpApplicationException.Builder =
        HttpApplicationException.Builder().apply {
            if (body != null) {
               message = body
            }

            statusCode = resp.code()
            error = resp.message()
            exception = source.javaClass.name
            timestamp = System.currentTimeMillis()
        }